crawl-ref/source/webserver/game_data/static/action_panel.js

```diff
@@ -305,51 +301,121 @@ function ($, comm, client, cr, enums, options, player, icons, gui, util,
                     hide_tooltip();
                     tooltip_timeout = setTimeout(function()
                     {
-                        show_tooltip(ev.pageX, ev.pageY, selected - 1);
+                        show_tooltip(ev.pageX, ev.pageY,
+                                     selected - NUM_RESERVED_BUTTONS);
                     }, 500);
                 }
             }
             else if (ev.type === "mousedown" && ev.which == 1)
             {
-                if (selected == 0)
+                if (selected == 0) // It should be available even in targeting mode
                     hide_panel();
                 else if (game.get_input_mode() == enums.mouse_mode.COMMAND
-                    && selected > 0 && selected < filtered_inv.length + 1)
+                         && selected == 1)
+                {
+                    comm.send_message("main_menu_action");
+                }
+                else if (game.get_input_mode() == enums.mouse_mode.COMMAND
+                         && selected >= NUM_RESERVED_BUTTONS
+                         && selected < filtered_inv.length + NUM_RESERVED_BUTTONS)
                 {
                     comm.send_message("inv_item_action",
-                                      {slot: filtered_inv[selected - 1].slot});
+                                      {slot: filtered_inv[selected - NUM_RESERVED_BUTTONS].slot});
                 }
             }
             else if (ev.type === "mousedown" && ev.which == 3)
             {
                 if (selected == 0) // right click on the x shows settings
                     show_settings(ev);
                 else if (game.get_input_mode() == enums.mouse_mode.COMMAND
-                    && selected > 0 && selected < filtered_inv.length + 1)
+                         && selected >= NUM_RESERVED_BUTTONS
+                         && selected < filtered_inv.length + NUM_RESERVED_BUTTONS)
                 {
                     comm.send_message("inv_item_describe",
-                                      {slot: filtered_inv[selected - 1].slot});
+                                      {slot: filtered_inv[selected - NUM_RESERVED_BUTTONS].slot});
                 }
             }
         }
     }
 
+    function draw_action(texture, tiles, offset, scale, needs_cursor, text)
+    {
+        tiles = Array.isArray(tiles) ? tiles : [tiles];
```

Comment:
maybe refactor to call the glyph drawing code from here instead of the main loop? Looks like the quantity/highlight code is duplicated right now unless I'm missing something